### ISSUE TITLE
The active model `where` method should act as a normal parameterized scope

### ISSUE THREAD

**Issue by catmando** [hyperstack-org/hyperstack#116]
Currently returns nil if used on a AR class, and throws a non-related error if used on a scope.

THIS WAS THE ORIGINAL CONCLUSION BUT ITS INCORRECT SEE DISCUSSION: 
> The reasons we won't implement where is that it could be used to insecurely determine counts on data.  
For example `Customers.where(top_secret_attribute: true).count` could be executed from the JS console to find the count of customers with the top_secret_attribute.

`where` should be implemented as a scope.   

We will need to add scope called `__hyperstack_internal_where_scope` 

```ruby
class ActiveRecord::Base 
  scope __hyperstack_internal_where_scope, 
    ->(attrs) { where(attrs) }, # server side we just call where
    filter: ->(attrs) { !attrs.detect { |k, v| self[k] != v } # client side optimization
  def self.where(attrs) 
    __hyperstack_internal_where_scope(attrs)
  end if RUBY_ENGINE == 'opal'
end
```

On the client side we can add the class level `where` method to `ActiveRecord::Base` and which simpy calls the `__hyperstack_internal_where_scope.

**Comment by Tim-Blokdijk** [comment]
where is working within scopes, I presume this is not a security issue?

Suggestion for an error message:
Hypermodel does not support `where` outside of a scope for "security reasons". See [link to explanation] for details.
Please use  `scopes`, `find`, `find_by` and ActiveRecord `relations` to query your data.
If you think you can write a secure `where` implementation, feel free to impress us.

**Comment by catmando** [comment]
@Tim-Blokdijk yes you are right, and in fact I am wrong in my overall conclusion.   `where` can can act like any other scope, and will simply following the normal security rules.  Thus if you do not open up the `all` scope then doing Customers.where(top_secret_attribute: true) will fail because no scope ever granted permission.

In practical terms you could do this:

```ruby
class Customer < ApplicationRecord
  regulate_scope :all, ->() { acting_user.admin? }  # admins can get any data needed
end
```

now

```ruby
Customer.where(top_secret_attribute: true).count
```

will only work if executed by an admin.

**Comment by catmando** [comment]
closed in 1.0.alpha1.7
